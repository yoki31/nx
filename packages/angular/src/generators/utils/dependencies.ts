import { addDependenciesToPackageJson, Tree } from '@nx/devkit';
import { versions } from './version-utils';

export function addBuildableLibrariesPostCssDependencies(tree: Tree): void {
  const pkgVersions = versions(tree);
  addDependenciesToPackageJson(
    tree,
    {},
    {
      postcss: pkgVersions.postcssVersion,
      autoprefixer: pkgVersions.autoprefixerVersion,
      'postcss-url': pkgVersions.postcssUrlVersion,
    }
  );
}
