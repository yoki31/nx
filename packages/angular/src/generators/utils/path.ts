import { joinPathFragments } from '@nx/devkit';
import { basename, dirname, relative } from 'path';

export function getRelativeImportToFile(
  sourceFilePath: string,
  targetFilePath: string
): string {
  const relativeDirToTarget = relative(
    dirname(sourceFilePath),
    dirname(targetFilePath)
  );

  return `./${joinPathFragments(
    relativeDirToTarget,
    basename(targetFilePath, '.ts')
  )}`;
}
