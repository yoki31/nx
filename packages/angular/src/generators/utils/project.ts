import type { Tree } from '@nx/devkit';
import { readProjectConfiguration } from '@nx/devkit';
import type { AngularProjectConfiguration } from '../../utils/types';

export function getProjectPrefix(
  tree: Tree,
  project: string
): string | undefined {
  return (
    readProjectConfiguration(tree, project) as AngularProjectConfiguration
  ).prefix;
}
