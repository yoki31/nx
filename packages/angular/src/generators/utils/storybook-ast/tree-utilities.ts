import type { Tree } from '@nx/devkit';
import { joinPathFragments } from '@nx/devkit';
import ignore from 'ignore';

export function getAllFilesRecursivelyFromDir(
  tree: Tree,
  dir: string
): string[] {
  if (isPathIgnored(tree, dir)) {
    return [];
  }

  const files: string[] = [];
  const children = tree.children(dir);
  children.forEach((child) => {
    const childPath = joinPathFragments(dir, child);
    if (tree.isFile(childPath)) {
      files.push(childPath);
    } else {
      files.push(...getAllFilesRecursivelyFromDir(tree, childPath));
    }
  });

  return files;
}

function isPathIgnored(tree: Tree, path: string): boolean {
  let ig: ReturnType<typeof ignore>;
  if (tree.exists('.gitignore')) {
    ig = ignore();
    ig.add(tree.read('.gitignore', 'utf-8'));
  }

  return path !== '' && ig?.ignores(path);
}
