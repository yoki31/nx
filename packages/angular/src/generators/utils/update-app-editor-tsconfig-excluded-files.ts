import {
  joinPathFragments,
  readJson,
  updateJson,
  type ProjectConfiguration,
  type Tree,
} from '@nx/devkit';

export function updateAppEditorTsConfigExcludedFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration
) {
  if (projectConfig.projectType !== 'application') {
    return;
  }

  const editorTsConfigPath = joinPathFragments(
    projectConfig.root,
    'tsconfig.editor.json'
  );
  const appTsConfigPath = joinPathFragments(
    projectConfig.root,
    'tsconfig.app.json'
  );
  if (!tree.exists(editorTsConfigPath) || !tree.exists(appTsConfigPath)) {
    return;
  }

  const appTsConfig = readJson(tree, appTsConfigPath);
  updateJson(tree, editorTsConfigPath, (json) => {
    const exclude = [...(json.exclude ?? []), ...(appTsConfig.exclude ?? [])];
    if (exclude.length) {
      json.exclude = Array.from(new Set(exclude));
    }

    return json;
  });
}
