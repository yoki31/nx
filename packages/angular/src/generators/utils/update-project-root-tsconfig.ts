import { updateJson, type Tree } from '@nx/devkit';
import { tsConfigBaseOptions } from '@nx/js';

export { extractTsConfigBase } from '@nx/js';

export function updateProjectRootTsConfig(
  host: Tree,
  projectRoot: string,
  relativePathToRootTsConfig: string,
  isRootProject?: boolean
) {
  updateJson(host, `${projectRoot}/tsconfig.json`, (json) => {
    if (isRootProject) {
      // inline tsconfig.base.json into the project
      json.compileOnSave = false;
      json.compilerOptions = {
        ...tsConfigBaseOptions,
        ...json.compilerOptions,
      };
      json.exclude = ['node_modules', 'tmp'];
      delete json.extends;
    } else {
      json.extends = relativePathToRootTsConfig;
    }

    return json;
  });
}
