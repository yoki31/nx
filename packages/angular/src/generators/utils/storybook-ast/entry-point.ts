import type { ProjectType, Tree } from '@nx/devkit';
import {
  joinPathFragments,
  normalizePath,
  readJson,
  readProjectConfiguration,
  visitNotIgnoredFiles,
} from '@nx/devkit';
import { basename, dirname } from 'path';

export type EntryPoint = { name: string; path: string; excludeDirs?: string[] };

export function getProjectEntryPoints(
  tree: Tree,
  projectName: string
): EntryPoint[] {
  const { root, sourceRoot, projectType } = readProjectConfiguration(
    tree,
    projectName
  );

  const rootEntryPoint: EntryPoint = {
    name: '',
    path: normalizeMainEntryPointSourceRoot(
      tree,
      root,
      sourceRoot,
      projectType
    ),
  };
  const entryPointRootPaths = [rootEntryPoint];

  if (projectType === 'application') {
    return entryPointRootPaths;
  }

  collectLibrarySecondaryEntryPoints(tree, root, entryPointRootPaths);

  // since the root includes some secondary entry points, we need to ignore
  // them when processing the main entry point
  if (rootEntryPoint.path === root && entryPointRootPaths.length > 1) {
    rootEntryPoint.excludeDirs = entryPointRootPaths
      .slice(1)
      .map((entryPoint) => entryPoint.path);
  }

  return entryPointRootPaths;
}

function collectLibrarySecondaryEntryPoints(
  tree: Tree,
  root: string,
  entryPointPaths: EntryPoint[]
): void {
  const exclude = new Set([`${root}/ng-package.json`, `${root}/package.json`]);

  visitNotIgnoredFiles(tree, root, (path) => {
    const normalizedPath = normalizePath(path);

    if (!tree.isFile(normalizedPath) || exclude.has(normalizedPath)) {
      return;
    }

    const fileName = basename(normalizedPath);
    if (
      fileName !== 'ng-package.json' &&
      (fileName !== 'package.json' ||
        (fileName === 'package.json' &&
          !readJson(tree, normalizedPath).ngPackage))
    ) {
      return;
    }

    const entryPointPath = getSourcePath(
      tree,
      normalizePath(dirname(normalizedPath)),
      'lib'
    );

    entryPointPaths.push({
      name: basename(dirname(normalizedPath)),
      path: entryPointPath,
    });
  });
}

function normalizeMainEntryPointSourceRoot(
  tree: Tree,
  root: string,
  sourceRoot: string,
  projectType: ProjectType
): string {
  const projectTypeDir = projectType === 'application' ? 'app' : 'lib';

  if (sourceRoot) {
    return [joinPathFragments(sourceRoot, projectTypeDir), sourceRoot].find(
      (path) => tree.exists(path)
    );
  }

  return getSourcePath(tree, root, projectTypeDir) ?? root;
}

function getSourcePath(
  tree: Tree,
  basePath: string,
  projectTypeDir: string
): string | undefined {
  const candidatePaths = [
    joinPathFragments(basePath, 'src', projectTypeDir),
    joinPathFragments(basePath, 'src'),
    joinPathFragments(basePath, projectTypeDir),
    basePath,
  ];

  return candidatePaths.find((candidatePath) => tree.exists(candidatePath));
}
