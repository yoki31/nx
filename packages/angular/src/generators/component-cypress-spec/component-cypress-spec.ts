import type { Tree } from '@nx/devkit';
import {
  formatFiles,
  generateFiles,
  joinPathFragments,
  readProjectConfiguration,
} from '@nx/devkit';
import { getComponentProps } from '../utils/storybook-ast/storybook-inputs';
import { getArgsDefaultValue } from './lib/get-args-default-value';
import { getComponentSelector } from './lib/get-component-selector';
import type { ComponentCypressSpecGeneratorOptions } from './schema';

export async function componentCypressSpecGenerator(
  tree: Tree,
  options: ComponentCypressSpecGeneratorOptions
): Promise<void> {
  const {
    cypressProject,
    projectName,
    projectPath,
    componentPath,
    componentFileName,
    componentName,
    specDirectory,
  } = options;
  const e2eProjectName = cypressProject || `${projectName}-e2e`;
  const { sourceRoot, root } = readProjectConfiguration(tree, e2eProjectName);
  const isCypressV10 = tree.exists(
    joinPathFragments(root, 'cypress.config.ts')
  );
  const e2eLibIntegrationFolderPath = joinPathFragments(
    sourceRoot,
    isCypressV10 ? 'e2e' : 'integration'
  );

  const templatesDir = joinPathFragments(__dirname, 'files');
  const destinationDir = joinPathFragments(
    e2eLibIntegrationFolderPath,
    specDirectory ?? componentPath
  );
  const storyFile = joinPathFragments(
    destinationDir,
    `${componentFileName}.${isCypressV10 ? 'cy' : 'spec'}.ts`
  );

  if (tree.exists(storyFile)) {
    return;
  }

  const fullComponentPath = joinPathFragments(
    projectPath,
    componentPath,
    `${componentFileName}.ts`
  );
  const props = getComponentProps(tree, fullComponentPath, getArgsDefaultValue);
  const componentSelector = getComponentSelector(tree, fullComponentPath);

  generateFiles(tree, templatesDir, destinationDir, {
    projectName,
    componentFileName,
    componentName,
    componentSelector,
    props,
    fileExt: isCypressV10 ? 'cy.ts' : 'spec.ts',
  });

  if (!options.skipFormat) {
    await formatFiles(tree);
  }
}

export default componentCypressSpecGenerator;
