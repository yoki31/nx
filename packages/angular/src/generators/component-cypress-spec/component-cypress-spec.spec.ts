import 'nx/src/internal-testing-utils/mock-project-graph';

import { installedCypressVersion } from '@nx/cypress/src/utils/cypress-version';
import type { Tree } from '@nx/devkit';
import * as devkit from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { componentGenerator } from '../component/component';
import * as storybookUtils from '../utils/storybook-ast/storybook-inputs';
import { generateTestApplication } from '../utils/testing';
import { componentCypressSpecGenerator } from './component-cypress-spec';
import { E2eTestRunner } from '../../utils/test-runners';

// need to mock cypress otherwise it'll use the nx installed version from package.json
//  which is v9 while we are testing for the new v10 version
jest.mock('@nx/cypress/src/utils/cypress-version');

describe('componentCypressSpec generator', () => {
  let tree: Tree;
  const appName = 'ng-app1';
  const specFile = `${appName}-e2e/src/e2e/test-button/test-button.component.cy.ts`;
  let mockedInstalledCypressVersion: jest.Mock<
    ReturnType<typeof installedCypressVersion>
  > = installedCypressVersion as never;
  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });

    await generateTestApplication(tree, {
      name: appName,
      skipFormat: true,
      e2eTestRunner: E2eTestRunner.Cypress,
    });
    await componentGenerator(tree, {
      name: 'test-button',
      project: appName,
      skipFormat: true,
    });

    tree.write(
      `${appName}/src/app/test-button/test-button.component.ts`,
      `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;      
}`
    );
  });

  it('should not generate the component spec file when it already exists', async () => {
    mockedInstalledCypressVersion.mockReturnValue(10);
    jest.spyOn(storybookUtils, 'getComponentProps');
    jest.spyOn(devkit, 'generateFiles');
    tree.write(specFile, '');

    await componentCypressSpecGenerator(tree, {
      componentFileName: 'test-button.component',
      componentName: 'TestButtonComponent',
      componentPath: `test-button`,
      projectPath: `${appName}/src/app`,
      projectName: appName,
      skipFormat: true,
    });

    expect(storybookUtils.getComponentProps).not.toHaveBeenCalled();
    expect(devkit.generateFiles).not.toHaveBeenCalled();
    expect(tree.read(specFile).toString()).toBe('');
  });

  it('should generate the component spec file', async () => {
    mockedInstalledCypressVersion.mockReturnValue(10);
    await componentCypressSpecGenerator(tree, {
      componentFileName: 'test-button.component',
      componentName: 'TestButtonComponent',
      componentPath: `test-button`,
      projectPath: `${appName}/src/app`,
      projectName: appName,
    });

    expect(tree.exists(specFile)).toBe(true);
    const specFileContent = tree.read(specFile).toString();
    expect(specFileContent).toMatchSnapshot();
  });

  it('should generate .spec.ts when using cypress.json', async () => {
    mockedInstalledCypressVersion.mockReturnValue(9);
    const v9SpecFile = `${appName}-e2e/src/integration/test-button/test-button.component.spec.ts`;
    tree.delete(`${appName}-e2e/cypress.config.ts`);
    tree.write(`${appName}-e2e/cypress.json`, `{}`);

    await componentCypressSpecGenerator(tree, {
      componentFileName: 'test-button.component',
      componentName: 'TestButtonComponent',
      componentPath: `test-button`,
      projectPath: `${appName}/src/app`,
      projectName: appName,
      skipFormat: true,
    });

    expect(tree.exists(v9SpecFile)).toBe(true);
    const specFileContent = tree.read(v9SpecFile).toString();
    expect(specFileContent).toMatchSnapshot();
  });
});
