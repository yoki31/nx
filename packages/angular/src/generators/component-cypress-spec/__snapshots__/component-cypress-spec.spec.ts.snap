// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componentCypressSpec generator should generate .spec.ts when using cypress.json 1`] = `
"describe('ng-app1', () => {
  beforeEach(() =>
    cy.visit(
      '/iframe.html?id=testbuttoncomponent--primary&args=buttonType:button;style:default;age;isOn:false;'
    )
  );

  it('should render the component', () => {
    cy.get('proj-test-button').should('exist');
  });
});
"
`;

exports[`componentCypressSpec generator should generate the component spec file 1`] = `
"describe('ng-app1', () => {
  beforeEach(() =>
    cy.visit(
      '/iframe.html?id=testbuttoncomponent--primary&args=buttonType:button;style:default;age;isOn:false;'
    )
  );

  it('should render the component', () => {
    cy.get('proj-test-button').should('exist');
  });
});
"
`;
